library(Seurat)
library(ArchR)
addArchRThreads(threads = 16)

#Read cattle genome data and reference files
library(BSgenome.Btaurus.UCSC.bosTau9)
library(TxDb.Btaurus.UCSC.bosTau9.refGene)
library(org.Bt.eg.db)
genomeAnnotation <- createGenomeAnnotation(genome = BSgenome.Btaurus.UCSC.bosTau9)
geneAnnotation <- createGeneAnnotation(TxDb = TxDb.Btaurus.UCSC.bosTau9.refGene, OrgDb = org.Bt.eg.db)

setwd("/home/Jingliangliang/SC")
tissue<-"all"
organ<-sort(c("brain","brain2","heart","heart2","kidney","kidney2","lung","lung2","muscle","muscle2","spleen0","spleen2","skin0","skin2","rumen","rumen2","fat0","fat2","liver0","liver2","ovary","testis"))#sample name
filesource<-"/home/Jingliangliang/SC"
filelist<-paste0("/home/Jingliangliang/SC/combine_",tissue,"/")

#Read raw data generated by cellranger-arc
dir.create(filelist, showWarnings = FALSE)
for(i in 1:length(organ))
{
	system(paste0("cp ",filesource,organ[i],"/atac_fragments.tsv.gz ",filelist,organ[i],".gz"))
	system(paste0("cp ",filesource,organ[i],"/atac_fragments.tsv.gz.tbi ",filelist,organ[i],".gz.tbi"))
	system(paste0("cp ",filesource,organ[i],"/filtered_feature_bc_matrix.h5 ",filelist,organ[i],".h5"))
}

#Create arrow file
inputFiles <- list.files(filelist, pattern = '.gz',full.names = TRUE)
names(inputFiles) <- gsub(".gz", "", list.files(filelist,pattern = ".gz"))
ArrowFiles <- createArrowFiles(inputFiles = inputFiles,sampleNames = names(inputFiles),minTSS = 2,minFrags = 500,addTileMat = TRUE,addGeneScoreMat = TRUE,geneAnnotation = geneAnnotation,genomeAnnotation = genomeAnnotation)
#ArrowFiles <- paste0(organ,".arrow")#If the arrow file has already been created

#Create ArchR project
proj<- ArchRProject(ArrowFiles = ArrowFiles, outputDirectory = paste0("proj_combine_",tissue),copyArrows = FALSE,geneAnnotation = geneAnnotation,genomeAnnotation = genomeAnnotation)


#Create summarized experiment for snRNA
#############################
library(tidyverse)
library(magrittr)
create_summarized_experiment <- function(organ) {
  dir.ls <- list.files(filelist, pattern = "\\.h5$", full.names = TRUE)
  names(dir.ls) <- organ
  for (i in seq_along(organ)) {
    assign(paste0("seRNA_", organ[i]), import10xFeatureMatrix(input = dir.ls[i], names = organ[i]))
  }
  if (length(organ) < 2) {stop("At least two organizations are required to merge.")}
  seRNA_genelist <- NULL
  for (i in seq_along(organ)) {
    matrix_name <- paste0("seRNA_", organ[i])
    matrix <- get(matrix_name)
    if (is.null(seRNA_genelist)) {
      seRNA_genelist <- rownames(matrix)
    } else {
      seRNA_genelist <- intersect(seRNA_genelist, rownames(matrix))
    }
  }
  if (length(seRNA_genelist) == 0) {stop("There were no overlapping genes")}
  combined_data <- NULL
  for (i in seq_along(organ)) {
    matrix_name <- paste0("seRNA_", organ[i])
    matrix <- get(matrix_name)
    matrix_data <- assays(matrix[seRNA_genelist, ])[[1]]
    if (is.null(combined_data)) {
      combined_data <- matrix_data
    } else {
      combined_data <- cbind(combined_data, matrix_data)
    }
  }
  seRNA <- SummarizedExperiment(
    assays = list(counts = combined_data),
    rowRanges = rowRanges(get(paste0("seRNA_", organ[1]))[seRNA_genelist, ])
  )
  return(seRNA)
}
seRNA <- create_summarized_experiment(organ)

#Add paired snRNA data to the ArchR project
proj <- addGeneExpressionMatrix(input = proj, seRNA = seRNA)

#This function is used to get low quality cell names
filter_low_quality_rna <- function(organ, mt.gene) {
  low_quality_cells <- list()
  for (i in seq_along(organ)) {
    file_path <- paste0(filelist, organ[i], ".h5")
    seurat_name <- paste0("seurat_", organ[i])
    rna_name <- paste0(seurat_name, "_RNA")
    assign(seurat_name, Read10X_h5(file_path))
    assign(rna_name, get(seurat_name)$`Gene Expression`)
	obj<-get(rna_name)
	colnames(obj) <- paste(organ[i], colnames(obj), sep = '#')
    seurat_object <- CreateSeuratObject(counts = obj)
    assign(organ[i], seurat_object)
    seurat_object[["percent.mt"]] <- PercentageFeatureSet(seurat_object, features = mt.gene)
    assign(organ[i], seurat_object)
    low_quality_cells[[organ[i]]] <- colnames(seurat_object)[seurat_object$nFeature_RNA < 200 | seurat_object$percent.mt > 5]
  }
  # Combine all low quality cells across organs
  low_quality_cells_combined <- unlist(low_quality_cells)
  return(list(
    seurat_objects = mget(organ),
    low_quality_cells = low_quality_cells_combined
  ))
}
#Mt genes in cattle
mt.gene <- c("ND1","ND2","COX1","COX2","ATP8","ATP6","COX3","ND3","ND4L","ND4","ND5","CYTB","ND6")
results <- filter_low_quality_rna(organ, mt.gene)
seurat_objects <- results$seurat_objects
low_quality_cells <- results$low_quality_cells
index1 <- (!proj$cellNames%in%low_quality_cells) & (!is.na(proj$Gex_nUMI))
proj <- subsetArchRProject(proj,cells=proj$cellNames[index1],outputDirectory = paste0("proj_combine_",tissue),force = TRUE)

#filter Doublets
proj <- addDoubletScores(proj)
proj <- filterDoublets(proj)

#######################
#clustree was used to test the clustering of snRNA data at different resolutions
library(cowplot)
library(DoubletFinder)
library(clustree)
library(harmony)
options(future.globals.maxSize = 1000 * 1024^2)
process_seurat_objects <- function(seurat_objects) {
  seurat_objects_transformed <- lapply(seurat_objects, function(obj) {
    SCTransform(obj, verbose = FALSE,return.only.var.genes=FALSE)
  })
  features <- SelectIntegrationFeatures(object.list = seurat_objects_transformed, nfeatures = 2000)
  seurat.combined <- Reduce(function(x, y) merge(x, y), seurat_objects_transformed)
  seurat.combined2 <- subset(seurat.combined, cells = proj$cellNames)
  sample <- subset(seurat.combined2, subset = nFeature_RNA > 200 & percent.mt < 5)
  sample <- RunPCA(sample, npcs = 30, features = features, verbose = FALSE)
  group_labels <- sub("#.*", "", colnames(sample))
  sample@meta.data$group <- group_labels
  sample <- sample %>% RunHarmony("group",plot_convergence = TRUE)
  return(sample)
}
sample <- process_seurat_objects(seurat_objects)
sample <- sample %>%
  RunUMAP(reduction = "harmony", dims = 1:30, n.neighbors = 10, min.dist = 0.5) %>%
  FindNeighbors(reduction = "harmony", dims = 1:30) %>%
  FindClusters(resolution = 0) %>%
  identity()
#DimPlot(sample, label = TRUE)
sample.final <- FindClusters(
    object = sample,
    resolution = c(seq(0,0.8,.1))
)
clustree<-clustree(sample.final@meta.data, prefix = "SCT_snn_res.")
clustree

#snMultiome dimension reduction
set.seed(1)
set_resolution <- 0.6
proj <-addIterativeLSI(ArchRProj = proj, 
    clusterParams = list(
      resolution = set_resolution, 
      sampleCells = 10000,
      n.start = 10
    ),
    saveIterations = FALSE,
    useMatrix = "TileMatrix", 
    depthCol = "nFrags",force = TRUE,
    name = paste0("LSI_ATAC_",set_resolution))
##RNA-LSI
proj <- addIterativeLSI(ArchRProj = proj,clusterParams = list(resolution = set_resolution,sampleCells = 10000,n.start = 10),saveIterations = FALSE,useMatrix = "GeneExpressionMatrix",depthCol = "Gex_nUMI",varFeatures = 2500,firstSelection = "variable",binarize = FALSE,force = TRUE,name = paste0("LSI_RNA_",set_resolution))
#combined
proj <- addCombinedDims(proj, reducedDims = c(paste0("LSI_ATAC_",set_resolution), paste0("LSI_RNA_",set_resolution)), name = "LSI_Combined")
#Harmony
proj$Clusters3 <- ifelse(grepl("2", proj$Sample), "Mongolian", "Leiqiong")
proj$tissue <- gsub("[0-9]", "", proj$Sample)
#proj$tissue <- gsub("fat", "Adipose", proj$tissue)
#Harmony and clustering
proj <- addHarmony(ArchRProj = proj,reducedDims = "LSI_Combined",name = "Harmony",groupBy = c("Sample"),lambda = 0.6,corCutOff = 0.7,force = TRUE)
proj <- addClusters(proj,reducedDims = "Harmony", method = "Seurat",name = "Clusters",resolution =set_resolution,maxClusters = NULL,force = TRUE)
proj <- addHarmony(ArchRProj = proj,reducedDims = paste0("LSI_RNA_",set_resolution),name = "Harmony_RNA",groupBy = c("Clusters3","Sample"),force = TRUE)
proj <- addHarmony(ArchRProj = proj,reducedDims = paste0("LSI_ATAC_",set_resolution),name = "Harmony_ATAC",groupBy = c("Clusters3","Sample"),force = TRUE)
proj <- addClusters(proj,reducedDims = "Harmony_ATAC", method = "Seurat",name = "Clusters_ATAC",resolution =set_resolution,maxClusters = NULL,force = TRUE)
proj <- addClusters(proj,reducedDims = "Harmony_RNA", method = "Seurat",name = "Clusters_RNA",resolution =set_resolution,maxClusters = NULL,force = TRUE)


proj <- addUMAP(ArchRProj = proj,reducedDims = "Harmony",name = "UMAP",minDist = 0.4,n_neighbors=100,metric ="euclidean",force = TRUE)

p1 <- plotEmbedding(proj, name = "Clusters", embedding = "UMAP", labelAsFactors=F,plotAs="points",size = 0.000000001, labelMeans=T,rastr = FALSE)
p2 <- plotEmbedding(proj, name = "tissue", embedding = "UMAP", labelAsFactors=F,plotAs="points",size = 0.000000001, labelMeans=F,rastr = FALSE)

saveRDS(proj, file = "combine_all_after_filter.rds")