library(Seurat)
library(Signac)
library(AnnotationHub)
library(ArchR)
library(SingleCellExperiment)
library(harmony)
addArchRThreads(threads = 16)

#Read cattle genome data and reference files
library(BSgenome.Btaurus.UCSC.bosTau9)
library(TxDb.Btaurus.UCSC.bosTau9.refGene)
library(org.Bt.eg.db)
genomeAnnotation <- createGenomeAnnotation(genome = BSgenome.Btaurus.UCSC.bosTau9)
geneAnnotation <- createGeneAnnotation(TxDb = TxDb.Btaurus.UCSC.bosTau9.refGene, OrgDb = org.Bt.eg.db)

setwd("/home/Jingliangliang/SC")
tissue<-"all"
organ<-sort(c("brain","brain2","heart","heart2","kidney","kidney2","lung","lung2","muscle","muscle2","spleen0","spleen2","skin0","skin2","rumen","rumen2","fat0","fat2","liver0","liver2","ovary","testis"))#sample name
filesource<-"/home/Jingliangliang/SC"
filelist<-paste0("/home/Jingliangliang/SC/combine_",tissue,"/")

#Read raw data generated by cellranger-arc
dir.create(filelist, showWarnings = FALSE)
for(i in 1:length(organ))
{
	system(paste0("cp ",filesource,organ[i],"/atac_fragments.tsv.gz ",filelist,organ[i],".gz"))
	system(paste0("cp ",filesource,organ[i],"/atac_fragments.tsv.gz.tbi ",filelist,organ[i],".gz.tbi"))
	system(paste0("cp ",filesource,organ[i],"/filtered_feature_bc_matrix.h5 ",filelist,organ[i],".h5"))
}

#Create arrow file
inputFiles <- list.files(filelist, pattern = '.gz',full.names = TRUE)
names(inputFiles) <- gsub(".gz", "", list.files(filelist,pattern = ".gz"))
ArrowFiles <- createArrowFiles(inputFiles = inputFiles,sampleNames = names(inputFiles),minTSS = 2,minFrags = 500,addTileMat = TRUE,addGeneScoreMat = TRUE,geneAnnotation = geneAnnotation,genomeAnnotation = genomeAnnotation)
#ArrowFiles <- paste0(organ,".arrow")#If the arrow file has already been created

#Create ArchR project
proj<- ArchRProject(ArrowFiles = ArrowFiles, outputDirectory = paste0("proj_combine_",tissue),copyArrows = FALSE,geneAnnotation = geneAnnotation,genomeAnnotation = genomeAnnotation)


#Create summarized experiment for snRNA
#############################
library(tidyverse)
library(magrittr)
create_summarized_experiment <- function(organ) {
  dir.ls <- list.files(filelist, pattern = "\\.h5$", full.names = TRUE)
  names(dir.ls) <- organ
  for (i in seq_along(organ)) {
    assign(paste0("seRNA_", organ[i]), import10xFeatureMatrix(input = dir.ls[i], names = organ[i]))
  }
  if (length(organ) < 2) {stop("At least two organizations are required to merge.")}
  seRNA_genelist <- NULL
  for (i in seq_along(organ)) {
    matrix_name <- paste0("seRNA_", organ[i])
    matrix <- get(matrix_name)
    if (is.null(seRNA_genelist)) {
      seRNA_genelist <- rownames(matrix)
    } else {
      seRNA_genelist <- intersect(seRNA_genelist, rownames(matrix))
    }
  }
  if (length(seRNA_genelist) == 0) {stop("There were no overlapping genes")}
  combined_data <- NULL
  for (i in seq_along(organ)) {
    matrix_name <- paste0("seRNA_", organ[i])
    matrix <- get(matrix_name)
    matrix_data <- assays(matrix[seRNA_genelist, ])[[1]]
    if (is.null(combined_data)) {
      combined_data <- matrix_data
    } else {
      combined_data <- cbind(combined_data, matrix_data)
    }
  }
  seRNA <- SummarizedExperiment(
    assays = list(counts = combined_data),
    rowRanges = rowRanges(get(paste0("seRNA_", organ[1]))[seRNA_genelist, ])
  )
  return(seRNA)
}
seRNA <- create_summarized_experiment(organ)

#Add paired snRNA data to the ArchR project
proj <- addGeneExpressionMatrix(input = proj, seRNA = seRNA)

#This function is used to get low quality cell names
filter_low_quality_rna <- function(organ, mt.gene) {
  low_quality_cells <- list()
  for (i in seq_along(organ)) {
    file_path <- paste0(filelist, organ[i], ".h5")
    seurat_name <- paste0("seurat_", organ[i])
    rna_name <- paste0(seurat_name, "_RNA")
    assign(seurat_name, Read10X_h5(file_path))
    assign(rna_name, get(seurat_name)$`Gene Expression`)
	obj<-get(rna_name)
	colnames(obj) <- paste(organ[i], colnames(obj), sep = '#')
    seurat_object <- CreateSeuratObject(counts = obj)
    assign(organ[i], seurat_object)
    seurat_object[["percent.mt"]] <- PercentageFeatureSet(seurat_object, features = mt.gene)
    assign(organ[i], seurat_object)
    low_quality_cells[[organ[i]]] <- colnames(seurat_object)[seurat_object$nFeature_RNA < 200 | seurat_object$percent.mt > 5]
  }
  # Combine all low quality cells across organs
  low_quality_cells_combined <- unlist(low_quality_cells)
  return(list(
    seurat_objects = mget(organ),
    low_quality_cells = low_quality_cells_combined
  ))
}
#Mt genes in cattle
mt.gene <- c("ND1","ND2","COX1","COX2","ATP8","ATP6","COX3","ND3","ND4L","ND4","ND5","CYTB","ND6")
results <- filter_low_quality_rna(organ, mt.gene)
seurat_objects <- results$seurat_objects
low_quality_cells <- results$low_quality_cells
index1 <- (!proj$cellNames%in%low_quality_cells) & (!is.na(proj$Gex_nUMI))
proj <- subsetArchRProject(proj,cells=proj$cellNames[index1],outputDirectory = paste0("proj_combine_",tissue),force = TRUE)

#filter Doublets
proj <- addDoubletScores(proj)
proj <- filterDoublets(proj)

#######################
#snMultiome dimension reduction
set.seed(1)
set_resolution <- 0.8
proj <-addIterativeLSI(ArchRProj = proj, 
    clusterParams = list(
      resolution = set_resolution, 
      sampleCells = 10000,
      n.start = 10
    ),
    saveIterations = FALSE,
    useMatrix = "TileMatrix", 
    depthCol = "nFrags",force = TRUE,
    name = paste0("LSI_ATAC_",set_resolution))

proj <- addHarmony(ArchRProj = proj,reducedDims = paste0("LSI_ATAC_",set_resolution),name = "Harmony_ATAC",groupBy = c("Clusters3","Sample"),force = TRUE)

proj$Clusters3 <- ifelse(grepl("2", proj$Sample), "Mongolian", "Leiqiong")
proj$tissue <- gsub("[0-9]", "", proj$Sample)
#proj$tissue <- gsub("fat", "Adipose", proj$tissue)

##RNA-LSI
##!Due to the limitations of LSI, RNA dimensionality reduction was changed to be conducted in Seurat.
##After ArchR2Seurat (see Data_format_conversion.R)
DefaultAssay(SeuratObject) <- "RNA"
SeuratObject <- SCTransform(SeuratObject, verbose = FALSE) %>% 
	RunPCA() %>% 
	RunHarmony(group.by.vars = "Sample", reduction = "pca", reduction.save = "harmony_rna",dims.use = 1:30) %>% 
	RunUMAP(dims = 1:30, reduction = "harmony_rna", reduction.name = 'umap.rna', reduction.key = 'rnaUMAP_')

colnames(SeuratObject) <- sub("_", "#", colnames(SeuratObject))
LSI_ATAC<- getReducedDims(proj, reducedDims ="Harmony_ATAC", returnMatrix = TRUE)

if(all(rownames(LSI_ATAC) %in% colnames(SeuratObject))){
	SeuratObject[["harmony_atac"]] <- CreateDimReducObject(
		embeddings = as.matrix(LSI_ATAC),
		key = "LSIATAC_",
		assay = "peaks"
	)}

SeuratObject <- FindMultiModalNeighbors(SeuratObject, reduction.list = list("harmony_rna", "harmony_atac"), dims.list = list(1:30, 2:30))
SeuratObject <- RunUMAP(SeuratObject, nn.name = "weighted.nn", reduction.name = "wnn.umap", reduction.key = "wnnUMAP_")
SeuratObject <- FindClusters(SeuratObject, graph.name = "wsnn", algorithm = 3, verbose = FALSE)
#Add the Suerat data to the Archr project
#reduction
harmony_rna_mat <- Embeddings(SeuratObject, reduction = "harmony_rna")[, 1:30]
harmony_rna_mat <- harmony_rna_mat[proj$cellNames, ]
rna_reducedDims <- proj@reducedDims$Harmony_ATAC
rna_reducedDims@listData$matDR <- harmony_rna_mat
rna_reducedDims@listData$date <- Sys.time()
proj@reducedDims$harmony_rna <- rna_reducedDims
proj <- addCombinedDims(proj, reducedDims = c("harmony_rna","Harmony_ATAC"), name = "Harmony")
#wnn.umap
seurat.umap <- SeuratObject@reductions$wnn.umap
seurat.umap.df <- DataFrame(row.names=proj$cellNames, "seurat#UMAP1" = seurat.umap@cell.embeddings[proj$cellNames, "wnnUMAP_1"], "seurat#UMAP2" =  seurat.umap@cell.embeddings[proj$cellNames, "wnnUMAP_2"], check.names = FALSE)
seurat.umap.df$cn <- rownames(seurat.umap.df)
ArchR.meta <- proj@cellColData
ArchR.meta$cn <- rownames(ArchR.meta)
ordered.coords <- merge(ArchR.meta, seurat.umap.df, by = "cn")[, c("seurat#UMAP1", "seurat#UMAP2", "cn")]
rownames(ordered.coords) <- ordered.coords$cn
ordered.coords$cn <- NULL
proj@embeddings$UMAP_wnn <- SimpleList(df = seurat.umap.df, params = list())
#clusters
wnn_clusters <- SeuratObject$seurat_clusters
wnn_clusters <- wnn_clusters[proj$cellNames]
proj$WNN_Clusters <- wnn_clusters

saveRDS(proj, file = "combine_all_after_filter.rds")