#setwd("/home/Jingliangliang/liftover/50")
setwd("/storage/public/home/2021060195/SC/liftover/50-merge-nonredundant/")
library(data.table)
for(j in c("Hg38","Mm10","SusScr11","EquCab3","GCF_016772045.2")){
  assign(paste("bos", j, sep = "_"), read.table(paste("bos","_0.5_", j,".bed", sep = ""),header=F))
  assign(paste("bos", j,"unmap", sep = "_"), read.table(paste("bos","_unmap_0.5_", j,".bed", sep = ""),header=F))
  assign(paste("bos", j,"unmap_set", sep = "_"), paste(get(paste("bos", j,"unmap", sep = "_"))[,1],get(paste("bos", j,"unmap", sep = "_"))[,2],get(paste("bos", j,"unmap", sep = "_"))[,3],sep = "-"))  
}

CRE_peak<-as.data.frame(fread("/storage/public/home/2021060195/SC/liftover/peakset_508551.csv",header=T))
CRE_set<-paste(CRE_peak[,1],CRE_peak[,2],CRE_peak[,3],sep = "-")#chr1-1-10 formal string
library(tidyverse)
set<-as.data.frame(list(name=CRE_set,Cattle="NULL",Level0="NULL",Level1="NULL",Level2="NULL",Level3="NULL"))
set$Level0<-!(CRE_peak$Leiqiong==1&CRE_peak$Mongolian==1)+0 #品种特有的CRE

#Note that the files generated by liftover have been converted to other species, the umap files should be used for matching.
set1 <- as.data.frame(list(name = CRE_set,hg = "NULL", mm = "NULL", pig = "NULL", horse = "NULL", sheep = "NULL"))
set1$hg[! CRE_set %in% bos_Hg38_unmap_set] <- "1"
set1$hg[CRE_set %in% bos_Hg38_unmap_set] <- "0"
set1$mm[! CRE_set %in% bos_Mm10_unmap_set] <- "1"
set1$mm[CRE_set %in% bos_Mm10_unmap_set] <- "0"
set1$pig[! CRE_set %in% bos_SusScr11_unmap_set] <- "1"
set1$pig[CRE_set %in% bos_SusScr11_unmap_set] <- "0"
set1$horse[! CRE_set %in% bos_EquCab3_unmap_set] <- "1"
set1$horse[CRE_set %in% bos_EquCab3_unmap_set] <- "0"
set1$sheep[! CRE_set %in% bos_GCF_016772045.2_unmap_set] <- "1"
set1$sheep[CRE_set %in% bos_GCF_016772045.2_unmap_set] <- "0"
set1[2:ncol(set1)] <- lapply(set1[2:ncol(set1)], as.numeric)


upset_data<-set1
upset_data$unmap_hg <- (set1$hg==0)+0
upset_data$unmap_mm <- (set1$mm==0)+0
upset_data$unmap_pig <- (set1$pig==0)+0
upset_data$unmap_horse <- (set1$horse==0)+0
upset_data$unmap_sheep <- (set1$sheep==0)+0

library(UpSetR)
pdf("upsetplot.pdf", width=10, height=10)
upset(upset_data, nsets = 5, mb.ratio = c(0.6, 0.4),sets=c("sheep","horse","pig","mm","hg"),keep.order = TRUE,order.by = c("freq", "degree"), decreasing = c(TRUE,TRUE),sets.bar.color = c("#20498D", "#20498D", "#20498D", "#20498D","#20498D"),matrix.color = '#225EA8',point.size = 3.3,line.size = 0.8,shade.color = 'grey',shade.alpha = 0.2,matrix.dot.alpha = 0.7)
dev.off()



cattle_specific0<-unique(Reduce(intersect,list(HN_Hg38_unmap_set,HN_Mm10_unmap_set,HN_SusScr11_unmap_set,HN_EquCab3_unmap_set,HN_GCF_016772045.2_unmap_set)),Reduce(intersect,list(MG_Hg38_unmap_set,MG_Mm10_unmap_set,MG_SusScr11_unmap_set,MG_EquCab3_unmap_set,MG_GCF_016772045.2_unmap_set)))
#############################################################
TSS<-read.table("/home/Jingliangliang/SC/gene_TSSup1k_down1k.txt",header=F)
TSS$V1<-paste0("chr",TSS$V1)
all_info_separated<-rbind(HN_all_info_separated,MG_all_info_separated)

TSS_int<-bt.intersect(a = all_info_separated, b = TSS,wa=T,wb=T)
###distal赋值0,启动子近端有交叠赋值为1
set2$TSS<-TSS_int$V14
set2$TSS[which(set2$TSS>1)]<-1
set2$fold_q<-10^(-TSS_int$V9)

##哺乳动物
##set2$hg==1&set2$mm==1&set2$pig==1&set2$horse==1&set2$sheep==1,n=2336261
mam_Level1<-set2[set2$hg==1&set2$mm==1&set2$pig==1&set2$horse==1&set2$sheep==1&set2$Level1==1,]
mam_Level1$n<-1:nrow(mam_Level1)#1783137
mam_Level2<-set2[set2$hg==1&set2$mm==1&set2$pig==1&set2$horse==1&set2$sheep==1&set2$Level2==1,]
mam_Level2$n<-1:nrow(mam_Level2)#1481795
mam_Level3<-set2[set2$hg==1&set2$mm==1&set2$pig==1&set2$horse==1&set2$sheep==1&set2$Level3==1,]
mam_Level3$n<-1:nrow(mam_Level3)#1335464
mam_Levelx<-set2[set2$hg==1&set2$mm==1&set2$pig==1&set2$horse==1&set2$sheep==1&set2$Levelx==1,]
mam_Levelx$n<-1:nrow(mam_Levelx)#139530

##牛特有的
##set2$hg==0&set2$mm==0&set2$pig==0&set2$horse==0&set2$sheep==0,n=36999
Level1<-set2[set2$hg==0&set2$mm==0&set2$pig==0&set2$horse==0&set2$sheep==0&set2$Level1==1,]
Level1$n<-1:nrow(Level1)#21235
Level2<-set2[set2$hg==0&set2$mm==0&set2$pig==0&set2$horse==0&set2$sheep==0&set2$Level2==1,]
Level2$n<-1:nrow(Level2)#17676
Level3<-set2[set2$hg==0&set2$mm==0&set2$pig==0&set2$horse==0&set2$sheep==0&set2$Level3==1,]
Level3$n<-1:nrow(Level3)#15108
Levelx<-set2[set2$hg==0&set2$mm==0&set2$pig==0&set2$horse==0&set2$sheep==0&set2$Levelx==1,]
Levelx$n<-1:nrow(Levelx)#7773

##家畜动物 Livestock
##set2$hg==0&set2$mm==0&set2$pig==1&set2$horse==1&set2$sheep==1,n=170903
livestock_Level1<-set2[set2$hg==0&set2$mm==0&set2$pig==1&set2$horse==1&set2$sheep==1&set2$Level1==1,]
livestock_Level1$n<-1:nrow(livestock_Level1)#106571
livestock_Level2<-set2[set2$hg==0&set2$mm==0&set2$pig==1&set2$horse==1&set2$sheep==1&set2$Level2==1,]
livestock_Level2$n<-1:nrow(livestock_Level2)#83645
livestock_Level3<-set2[set2$hg==0&set2$mm==0&set2$pig==1&set2$horse==1&set2$sheep==1&set2$Level3==1,]
livestock_Level3$n<-1:nrow(livestock_Level3)#67195
livestock_Levelx<-set2[set2$hg==0&set2$mm==0&set2$pig==1&set2$horse==1&set2$sheep==1&set2$Levelx==1,]
livestock_Levelx$n<-1:nrow(livestock_Levelx)#24984
##牛科动物 bovidae
##set2$hg==0&set2$mm==0&set2$pig==0&set2$horse==0&set2$sheep==1,n=168108
bovidae_Level1<-set2[set2$hg==0&set2$mm==0&set2$pig==0&set2$horse==0&set2$sheep==1&set2$Level1==1,]
bovidae_Level1$n<-1:nrow(bovidae_Level1)#87423
bovidae_Level2<-set2[set2$hg==0&set2$mm==0&set2$pig==0&set2$horse==0&set2$sheep==1&set2$Level2==1,]
bovidae_Level2$n<-1:nrow(bovidae_Level2)#66139
bovidae_Level3<-set2[set2$hg==0&set2$mm==0&set2$pig==0&set2$horse==0&set2$sheep==1&set2$Level3==1,]
bovidae_Level3$n<-1:nrow(bovidae_Level3)#48557
bovidae_Levelx<-set2[set2$hg==0&set2$mm==0&set2$pig==0&set2$horse==0&set2$sheep==1&set2$Levelx==1,]
bovidae_Levelx$n<-1:nrow(bovidae_Levelx)#43428

# 创建一个2x2的画布布局
par(mfrow = c(2, 2))
pie(table(mam_Levelx$TSS), labels = names(table(mam_Levelx$TSS)), main = "mam_Levelx", col = rainbow(length(table(mam_Levelx$TSS))))
pie(table(mam_Level1$TSS), labels = names(table(mam_Level1$TSS)), main = "mam_Level1", col = rainbow(length(table(mam_Level1$TSS))))
pie(table(mam_Level2$TSS), labels = names(table(mam_Level2$TSS)), main = "mam_Level2", col = rainbow(length(table(mam_Level2$TSS))))
pie(table(mam_Level3$TSS), labels = names(table(mam_Level3$TSS)), main = "mam_Level3", col = rainbow(length(table(mam_Level3$TSS))))
##pie1.pdf
par(mfrow = c(2, 2))
pie(table(Levelx$TSS), labels = names(table(Levelx$TSS)), main = "Levelx", col = rainbow(length(table(Levelx$TSS))))
pie(table(Level1$TSS), labels = names(table(Level1$TSS)), main = "Level1", col = rainbow(length(table(Level1$TSS))))
pie(table(Level2$TSS), labels = names(table(Level2$TSS)), main = "Level2", col = rainbow(length(table(Level2$TSS))))
pie(table(Level3$TSS), labels = names(table(Level3$TSS)), main = "Level3", col = rainbow(length(table(Level3$TSS))))
##pie2.pdf
par(mfrow = c(2, 2))
pie(table(livestock_Levelx$TSS), labels = names(table(livestock_Levelx$TSS)), main = "livestock_Levelx", col = rainbow(length(table(livestock_Levelx$TSS))))
pie(table(livestock_Level1$TSS), labels = names(table(livestock_Level1$TSS)), main = "livestock_Level1", col = rainbow(length(table(livestock_Level1$TSS))))
pie(table(livestock_Level2$TSS), labels = names(table(livestock_Level2$TSS)), main = "livestock_Level2", col = rainbow(length(table(livestock_Level2$TSS))))
pie(table(livestock_Level3$TSS), labels = names(table(livestock_Level3$TSS)), main = "livestock_Level3", col = rainbow(length(table(livestock_Level3$TSS))))
##pie3.pdf

par(mfrow = c(2, 2))
pie(table(bovidae$TSS), labels = names(table(bovidae$TSS)), main = "bovidae", col = rainbow(length(table(bovidae$TSS))))
pie(table(livestock_Level1$TSS), labels = names(table(livestock_Level1$TSS)), main = "livestock_Level1", col = rainbow(length(table(livestock_Level1$TSS))))
pie(table(livestock_Level2$TSS), labels = names(table(livestock_Level2$TSS)), main = "livestock_Level2", col = rainbow(length(table(livestock_Level2$TSS))))
pie(table(livestock_Level3$TSS), labels = names(table(livestock_Level3$TSS)), main = "livestock_Level3", col = rainbow(length(table(livestock_Level3$TSS))))
##pie4.pdf


###################################phastCons打分
#BiocManager::install("phastCons100way.UCSC.hg38")
library(GenomicRanges)
library(phastCons100way.UCSC.hg38)
phast <- phastCons100way.UCSC.hg38

hg_index<-set1$Hg==1
mm_index<-set1$mm==1

#####在set1中写入liftover转换后的序列
set1[set2$Cattle==1&set1$Hg==1,9:11]<-HN_Hg38
set1[set2$Cattle==0&set1$Hg==1,9:11]<-MG_Hg38

set1[set2$Cattle==1&set1$mm==1,12:14]<-HN_mm
set1[set2$Cattle==0&set1$mm==1,12:14]<-MG_mm

mam_index<-(set2$hg==1&set2$mm==1&set2$pig==1&set2$horse==1&set2$sheep==1)
mam<-paste(set1$V1[mam_index],paste(set1$V2[mam_index],set1$V3[mam_index],sep="-"),sep=":")

GRanges_mam<-GRanges(mam)
mam_score<-gscores(phast, GRanges_mam)

set2$pileup<-c(HN_all_info$V6,MG_all_info$V6)
set2$enrichment<-c(HN_all_info$V9,MG_all_info$V9)

set_mm<-set2[mam_index,]
set_mm$mam_score<-mam_score$default


mam_liftover_Level1<-set_mm[set_mm$Level1==1,]
mam_liftover_Level1$n<-1:nrow(mam_liftover_Level1)
mam_liftover_Level2<-set_mm[set_mm$Level2==1,]
mam_liftover_Level2$n<-1:nrow(mam_liftover_Level2)
mam_liftover_Level3<-set_mm[set_mm$Level3==1,]
mam_liftover_Level3$n<-1:nrow(mam_liftover_Level3)
mam_liftover_Levelx<-set_mm[set_mm$Levelx==1,]
mam_liftover_Levelx$n<-1:nrow(mam_liftover_Levelx)

library(BioQC)
entropySpecificity(mam_liftover_Level1[,12], norm=F)
library(ggplot2)
library(patchwork)
d0<-data.frame(mam_liftover_Levelx$mam_score )
d1<-data.frame(mam_liftover_Level1$mam_score )
d2<-data.frame(mam_liftover_Level2$mam_score )
d3<-data.frame(mam_liftover_Level3$mam_score )
colnames(d0)<-"score"
colnames(d1)<-"score"
colnames(d2)<-"score"
colnames(d3)<-"score"
g0<-ggplot(d0, aes(x = score))+geom_density(fill = "yellow",adjust = 0.5)+ theme(panel.background = element_rect(fill = "white"))
g1<-ggplot(d1, aes(x = score))+geom_density(fill = "yellow",adjust = 0.5)+ theme(panel.background = element_rect(fill = "white"))+ylim(0, 3)
g2<-ggplot(d2, aes(x = score))+geom_density(fill = "yellow",adjust = 0.5)+ theme(panel.background = element_rect(fill = "white"))+ylim(0, 3)
g3<-ggplot(d3, aes(x = score))+geom_density(fill = "yellow",adjust = 0.5)+ theme(panel.background = element_rect(fill = "white"))+ylim(0, 3)
g0+g1+g2+g3
##Density
save(list=ls(),file="fig4-new.RData")

###################################phyloP打分
mam2<-set1[mam_index, c("V1", "V2","V3")]
write.table(mam2,file="/storage/public/home/2021060195/02.phyloP/region_file.bed",sep="\t",col.names = F,row.names = F,quote = FALSE)
##################
setwd("/storage/public/home/2021060195/02.phyloP")
library(data.table)
chr_files <- list.files(pattern = "^chr.*\\.bed$")
bed_data_list <- lapply(chr_files, function(file) {
fread(file, header = FALSE, col.names = c('chr', 'start', 'end', 'score'))
})
names(bed_data_list) <- chr_files
region_data <- fread('/storage/public/home/2021060195/02.phyloP/region_file.bed', header = FALSE, col.names = c('chr', 'start', 'end'))
library(parallel)
num_cores <- detectCores()
process_region <- function(i) {
region_chr <- region_data$chr[i]
region_start <- region_data$start[i]
region_end <- region_data$end[i]
match_chr <- match(paste0(region_chr, ".bed"), chr_files)
bed_data <- bed_data_list[[match_chr]]
scores <- bed_data[start >= region_start & end <= region_end, "score"]
return(scores)
}
region_scores <- mclapply(seq_len(nrow(region_data)), process_region, mc.cores = num_cores)
means <- sapply(region_scores, function(df) mean(df$score, na.rm = TRUE))
write.table(means,file="phyloP_means.txt",sep="\t",col.names = F,row.names = F,quote = FALSE)
######################


####################HN特异SNP list
library(data.table)
fimo_HN <- fread("fimo_HN.tsv",fill=TRUE)
fimo_HN<-fimo_HN[fimo_HN$`q-value`<0.05,]
#########
{##########HN
#HN_SNP<-read.table("HN.1freq.bed",header=F)
HN_SNP<-read.table("HN.0.3freq.bed",header=F)
HN_SNP$V1<-paste0("chr",HN_SNP$V1)
#all_info_separated[k,]			set2[k,]
j<-c()
for(i in 1:nrow(HN_SNP)){
k<-which((all_info_separated[,1]==HN_SNP[i,1])&(all_info_separated[,2]<=HN_SNP[i,2])&(all_info_separated[,3]>=HN_SNP[i,2]))
if(length(k)>0){
print(i);
j<-c(j,i)
}
}
HN_SNP2<-HN_SNP[j,]#在cCRE中的SNP
fimo_HN_SNP<-NULL
for(i in 1:nrow(HN_SNP2)){
inter_HN<-all_info_separated[which((all_info_separated[,1]==HN_SNP2[i,1])&(all_info_separated[,2]<=HN_SNP2[i,2])&(all_info_separated[,3]>=HN_SNP2[i,2])),]
fimo_inrer<-NULL
for(j in 1:nrow(inter_HN)){
pos<-paste0("::",inter_HN[j,1],":",inter_HN[j,2],"-",inter_HN[j,3])
fimo_inrer2<-fimo_HN[which(fimo_HN[,3]==pos),]
fimo_inrer2$pos<-paste0(HN_SNP2[i,1],":",HN_SNP2[i,2])
fimo_inrer<-rbind(fimo_inrer,fimo_inrer2)
}
fimo_HN_SNP<-rbind(fimo_HN_SNP,fimo_inrer)
print(paste0(i,"/",nrow(HN_SNP2)))
}
x1<-unlist(strsplit(fimo_HN_SNP$sequence_name, "::|:|-"))
x2<-unlist(strsplit(fimo_HN_SNP$pos, "::|:|-"))
fimo_HN_SNP$chr1<-x1[seq(1, length(x1), by = 4)+1]
fimo_HN_SNP$start1<-as.numeric(x1[seq(1, length(x1), by = 4)+2])+as.numeric(fimo_HN_SNP$start)
fimo_HN_SNP$stop1<-as.numeric(x1[seq(1, length(x1), by = 4)+2])+as.numeric(fimo_HN_SNP$stop)
fimo_HN_SNP$pos2<-x2[seq(1, length(x2), by = 2)]
fimo_HN_SNP$pos3<-x2[seq(1, length(x2), by = 2)+1]
fimo_HN_SNP$pd<-ifelse((fimo_HN_SNP$pos2==fimo_HN_SNP$chr1)&(fimo_HN_SNP$pos3>fimo_HN_SNP$start1)&(fimo_HN_SNP$pos3<fimo_HN_SNP$stop1),"Y","N")
####fimo_HN_SNP_freq1<-fimo_HN_SNP

###fimo_HN_SNP[fimo_HN_SNP$pd=="Y",]
fimo_HN_SNP_x3<-fimo_HN_SNP[fimo_HN_SNP$pd=="Y",] ##海南牛高频SNP在fimo有TF富集的条目
}
{##########MG
fimo_MG <- fread("fimo_MG.tsv",fill=TRUE)
fimo_MG<-fimo_MG[fimo_MG$`q-value`<0.05,]
MG_SNP<-read.table("MG.0.3freq.bed",header=F)
MG_SNP$V1<-paste0("chr",MG_SNP$V1)
#all_info_separated[k,]			set2[k,]
j<-c()
for(i in 1:nrow(MG_SNP)){
k<-which((all_info_separated[,1]==MG_SNP[i,1])&(all_info_separated[,2]<=MG_SNP[i,2])&(all_info_separated[,3]>=MG_SNP[i,2]))
if(length(k)>0){
print(paste0(i,"/",nrow(MG_SNP)));
j<-c(j,i)
}
}
MG_SNP2<-MG_SNP[j,]#在cCRE中的SNP
fimo_MG_SNP<-NULL
for(i in 1:nrow(MG_SNP2)){
inter_MG<-all_info_separated[which((all_info_separated[,1]==MG_SNP2[i,1])&(all_info_separated[,2]<=MG_SNP2[i,2])&(all_info_separated[,3]>=MG_SNP2[i,2])),]
fimo_inrer<-NULL
for(j in 1:nrow(inter_MG)){
pos<-paste0("::",inter_MG[j,1],":",inter_MG[j,2],"-",inter_MG[j,3])
fimo_inrer2<-fimo_MG[which(fimo_MG[,3]==pos),]
fimo_inrer2$pos<-paste0(MG_SNP2[i,1],":",MG_SNP2[i,2])
fimo_inrer<-rbind(fimo_inrer,fimo_inrer2)
}
fimo_MG_SNP<-rbind(fimo_MG_SNP,fimo_inrer)
print(paste0(i,"/",nrow(MG_SNP2)))
}
y1<-unlist(strsplit(fimo_MG_SNP$sequence_name, "::|:|-"))
y2<-unlist(strsplit(fimo_MG_SNP$pos, "::|:|-"))
fimo_MG_SNP$chr1<-y1[seq(1, length(y1), by = 4)+1]
fimo_MG_SNP$start1<-as.numeric(y1[seq(1, length(y1), by = 4)+2])+as.numeric(fimo_MG_SNP$start)
fimo_MG_SNP$stop1<-as.numeric(y1[seq(1, length(y1), by = 4)+2])+as.numeric(fimo_MG_SNP$stop)
fimo_MG_SNP$pos2<-y2[seq(1, length(y2), by = 2)]
fimo_MG_SNP$pos3<-y2[seq(1, length(y2), by = 2)+1]
fimo_MG_SNP$pd<-ifelse((fimo_MG_SNP$pos2==fimo_MG_SNP$chr1)&(fimo_MG_SNP$pos3>fimo_MG_SNP$start1)&(fimo_MG_SNP$pos3<fimo_MG_SNP$stop1),"Y","N")
####fimo_MG_SNP_freq1<-fimo_MG_SNP

###fimo_MG_SNP[fimo_MG_SNP$pd=="Y",]
fimo_MG_SNP_y3<-fimo_MG_SNP[fimo_MG_SNP$pd=="Y",] ##海南牛高频SNP在fimo有TF富集的条目
}
#############
##所有TF 统计
library(data.table)
data1 <- fread("fimo_MG.tsv",fill=TRUE)
data1<-data1[data1$`q-value`<0.05,]
data2 <- fread("fimo_HN.tsv",fill=TRUE)
data2<-data2[data2$`q-value`<0.05,]
#########
set_all <- cbind(all_info_separated[,c(1:3)],set2)
set_all$pos<-paste0("::",set_all$V1,":",set_all$V2,"-",set_all$V3)
set2_HN<-set_all[set_all$Cattle==1,]
set2_MG<-set_all[set_all$Cattle==0,]
data_HN<-cbind(data2,set2_HN[match(data2$sequence_name,set2_HN$pos),])
data_MG<-cbind(data1,set2_MG[match(data1$sequence_name,set2_MG$pos),])
####level3
data_all<-rbind(data_HN,data_MG)
sort(table(data_all$motif_alt_id[data_all$TSS==0&data_all$Level3==1]))
####Levelx
sort(table(data_all$motif_alt_id[data_all$TSS==1&data_all$Levelx==1]))
####
t2<-as.data.frame(table(data_HN$motif_alt_id,data_HN$TSS))
t2_0<-t2[t2$Var2==0,]#HN远端TSS
t2_1<-t2[t2$Var2==1,]#HN近端TSS
t22<-full_join(t2_0, t2_1, by = "Var1")
t22$sum<-t22$Freq.x+t22$Freq.y

t1<-as.data.frame(table(data_MG$motif_alt_id,data_MG$TSS))
t1_0<-t1[t1$Var2==0,]#MG远端TSS
t1_1<-t1[t1$Var2==1,]#MG近端TSS
t11<-full_join(t1_0, t1_1, by = "Var1")
t11$sum<-t11$Freq.x+t11$Freq.y

tt<-full_join(as.data.frame(table(data_MG$motif_alt_id)), as.data.frame(table(data_HN$motif_alt_id)), by = "Var1")
tt <- replace(tt, is.na(tt), 0)
tt$Var1[tt$Freq.x==0&tt$Freq.y>100]
tt$Var1[tt$Freq.y==0&tt$Freq.x>100]

ttt<-full_join(t11, t22, by = "Var1")
ttt <- replace(ttt, is.na(ttt), 0)
ttt$fisher<-NULL
ttt$fisher.est<-NULL
for(i in 1:nrow(ttt)){
fisher<-fisher.test(matrix(c(ttt[i,3]+ttt[i,8],sum(set2$TSS==0)-(ttt[i,3]+ttt[i,8]),ttt[i,5]+ttt[i,10],sum(set2$TSS==1)-(ttt[i,5]+ttt[i,10])),nrow=2))
ttt$fisher[i]<-fisher$p.value
ttt$fisher.est[i]<-fisher$estimate
}

motif_id motif_alt_id            sequence_name start stop strand   score  p-value q-value matched_sequence           pos chr1   start1    stop1 pos2     pos3 pd
1: MA1125.1       ZNF384 ::chr6:58499036-58499584   264  275      + 15.1489 8.84e-07   0.048     attaaaaaaaaa chr6:58499303 chr6 58499300 58499311 chr6 58499303  Y
#######################不同level的元件
levelx_fimo<-full_join(as.data.frame(sort(table(data_HN$motif_alt_id[data_HN$Levelx==1]))), as.data.frame(sort(table(data_MG$motif_alt_id[data_MG$Levelx==1]))), by = "Var1")
level1_fimo<-full_join(as.data.frame(sort(table(data_HN$motif_alt_id[data_HN$Level1==1]))), as.data.frame(sort(table(data_MG$motif_alt_id[data_MG$Level1==1]))), by = "Var1")
level2_fimo<-full_join(as.data.frame(sort(table(data_HN$motif_alt_id[data_HN$Level2==1]))), as.data.frame(sort(table(data_MG$motif_alt_id[data_MG$Level2==1]))), by = "Var1")
level3_fimo<-full_join(as.data.frame(sort(table(data_HN$motif_alt_id[data_HN$Level3==1]))), as.data.frame(sort(table(data_MG$motif_alt_id[data_MG$Level3==1]))), by = "Var1")
########################
zic3<-data_HN[data_HN$motif_alt_id=="Zic3",]
table(unlist(strsplit(zic3$name,"-"))[seq(1, nrow(zic3)*5, by = 5)])
table(unlist(strsplit(zic3$name,"-"))[seq(1, nrow(zic3)*5, by = 5)+1])
table(zic3$TSS)
zic3_tss1<-as.data.frame(table(unlist(strsplit(zic3$name[zic3$TSS=="1"],"-"))[seq(1, length(zic3$TSS=="1")*5, by = 5)+1]))
zic3_tss0<-as.data.frame(table(unlist(strsplit(zic3$name[zic3$TSS=="0"],"-"))[seq(1, length(zic3$TSS=="0")*5, by = 5)+1]))
full_join(zic3_tss1,zic3_tss0, by = "Var1")
table(unlist(strsplit(zic3$name,"-"))[seq(1, nrow(zic3), by = 5)+1])

#######CTCF
CTCF<-data_all[data_all$motif_alt_id=="CTCF",]

NR2F1<-data_HN[data_HN$motif_alt_id=="NR2F1",]
table(unlist(strsplit(NR2F1$name,"-"))[seq(1, nrow(NR2F1)*5, by = 5)])
table(unlist(strsplit(NR2F1$name,"-"))[seq(1, nrow(NR2F1)*5, by = 5)+1])
table(NR2F1$TSS)

NR6A1<-data_HN[data_HN$motif_alt_id=="NR6A1",]
table(unlist(strsplit(NR6A1$name,"-"))[seq(1, nrow(NR6A1)*5, by = 5)])
table(unlist(strsplit(NR6A1$name,"-"))[seq(1, nrow(NR6A1)*5, by = 5)+1])
table(NR6A1$TSS)

PKNOX1<-data_HN[data_HN$motif_alt_id=="PKNOX1",]
table(unlist(strsplit(PKNOX1$name,"-"))[seq(1, nrow(PKNOX1)*5, by = 5)])
table(unlist(strsplit(PKNOX1$name,"-"))[seq(1, nrow(PKNOX1)*5, by = 5)+1])
table(PKNOX1$TSS)

ELK4<-data_HN[data_HN$motif_alt_id=="ELK4",]
table(unlist(strsplit(ELK4$name,"-"))[seq(1, nrow(ELK4)*5, by = 5)])
table(unlist(strsplit(ELK4$name,"-"))[seq(1, nrow(ELK4)*5, by = 5)+1])
table(ELK4$TSS)
ELK4_tss1<-as.data.frame(table(unlist(strsplit(ELK4$name[ELK4$TSS=="1"],"-"))[seq(1, length(ELK4$TSS=="1")*5, by = 5)+1]))
ELK4_tss0<-as.data.frame(table(unlist(strsplit(ELK4$name[ELK4$TSS=="0"],"-"))[seq(1, length(ELK4$TSS=="0")*5, by = 5)+1]))
full_join(ELK4_tss1,ELK4_tss0, by = "Var1")

Stat6<-data_HN[data_HN$motif_alt_id=="Stat6",]
table(unlist(strsplit(Stat6$name,"-"))[seq(1, nrow(Stat6)*5, by = 5)])
table(unlist(strsplit(Stat6$name,"-"))[seq(1, nrow(Stat6)*5, by = 5)+1])
table(Stat6$TSS)

Msgn1<-data_HN[data_HN$motif_alt_id=="Msgn1",]
table(unlist(strsplit(Msgn1$name,"-"))[seq(1, nrow(Msgn1)*5, by = 5)])
table(unlist(strsplit(Msgn1$name,"-"))[seq(1, nrow(Msgn1)*5, by = 5)+1])
table(Msgn1$TSS)

Rarb<-data_HN[data_HN$motif_alt_id=="Rarb",]
table(unlist(strsplit(Rarb$name,"-"))[seq(1, nrow(Rarb)*5, by = 5)])
table(unlist(strsplit(Rarb$name,"-"))[seq(1, nrow(Rarb)*5, by = 5)+1])
table(Rarb$TSS)

TP73<-data_HN[data_HN$motif_alt_id=="TP73",]
table(unlist(strsplit(TP73$name,"-"))[seq(1, nrow(TP73)*5, by = 5)])
table(unlist(strsplit(TP73$name,"-"))[seq(1, nrow(TP73)*5, by = 5)+1])
table(TP73$TSS)

FOSL2<-data_HN[data_HN$motif_alt_id=="FOSL2::JUND",]
table(unlist(strsplit(FOSL2$name,"-"))[seq(1, nrow(FOSL2)*5, by = 5)])
table(unlist(strsplit(FOSL2$name,"-"))[seq(1, nrow(FOSL2)*5, by = 5)+1])
table(FOSL2$TSS)

ZNF675<-data_HN[data_HN$motif_alt_id=="ZNF675",]
table(unlist(strsplit(ZNF675$name,"-"))[seq(1, nrow(ZNF675)*5, by = 5)])
table(unlist(strsplit(ZNF675$name,"-"))[seq(1, nrow(ZNF675)*5, by = 5)+1])
table(ZNF675$TSS)
#####################
SP5<-data_MG[data_MG$motif_alt_id=="SP5",]
table(unlist(strsplit(SP5$name,"-"))[seq(1, nrow(SP5)*5, by = 5)])
table(unlist(strsplit(SP5$name,"-"))[seq(1, nrow(SP5)*5, by = 5)+1])
table(SP5$TSS)

ZNF382<-data_MG[data_MG$motif_alt_id=="ZNF382",]
table(unlist(strsplit(ZNF382$name,"-"))[seq(1, nrow(ZNF382)*5, by = 5)])
table(unlist(strsplit(ZNF382$name,"-"))[seq(1, nrow(ZNF382)*5, by = 5)+1])
table(ZNF382$TSS)

ATF7<-data_MG[data_MG$motif_alt_id=="ATF7",]
table(unlist(strsplit(ATF7$name,"-"))[seq(1, nrow(ATF7)*5, by = 5)])
table(unlist(strsplit(ATF7$name,"-"))[seq(1, nrow(ATF7)*5, by = 5)+1])
table(ATF7$TSS)

THRB<-data_MG[data_MG$motif_alt_id=="THRB",]
table(unlist(strsplit(THRB$name,"-"))[seq(1, nrow(THRB)*5, by = 5)])
table(unlist(strsplit(THRB$name,"-"))[seq(1, nrow(THRB)*5, by = 5)+1])
table(THRB$TSS)

NFIX<-data_MG[data_MG$motif_alt_id=="NFIX",]
table(unlist(strsplit(NFIX$name,"-"))[seq(1, nrow(NFIX)*5, by = 5)])
table(unlist(strsplit(NFIX$name,"-"))[seq(1, nrow(NFIX)*5, by = 5)+1])
table(NFIX$TSS)

MAFG<-data_MG[data_MG$motif_alt_id=="MAFG::NFE2L1",]
table(unlist(strsplit(MAFG$name,"-"))[seq(1, nrow(MAFG)*5, by = 5)])
table(unlist(strsplit(MAFG$name,"-"))[seq(1, nrow(MAFG)*5, by = 5)+1])
table(MAFG$TSS)

Mafg<-data_MG[data_MG$motif_alt_id=="Mafg",]
table(unlist(strsplit(Mafg$name,"-"))[seq(1, nrow(Mafg)*5, by = 5)])
table(unlist(strsplit(Mafg$name,"-"))[seq(1, nrow(Mafg)*5, by = 5)+1])
table(Mafg$TSS)

ETV2<-data_MG[data_MG$motif_alt_id=="ETV2::FIGLA",]
table(unlist(strsplit(ETV2$name,"-"))[seq(1, nrow(ETV2)*5, by = 5)])
table(unlist(strsplit(ETV2$name,"-"))[seq(1, nrow(ETV2)*5, by = 5)+1])
table(ETV2$TSS)

ZNF528<-data_MG[data_MG$motif_alt_id=="ZNF528",]
table(unlist(strsplit(ZNF528$name,"-"))[seq(1, nrow(ZNF528)*5, by = 5)])
table(unlist(strsplit(ZNF528$name,"-"))[seq(1, nrow(ZNF528)*5, by = 5)+1])
table(ZNF528$TSS)


pers<-cbind(data_all$enrichment,data_all$mam_score,data_all$motif_alt_id)[!is.na(data_all$mam_score),]


{##########HN-fst
HN_fst<-read.table("1950.select.fst0.5.SNP.bed",header=T)
HN_fst$CHROM<-paste0("chr",HN_fst$CHROM)
#all_info_separated[k,]			set2[k,]
j<-c()
for(i in 1:nrow(HN_fst)){
k<-which((all_info_separated[,1]==HN_fst[i,1])&(all_info_separated[,2]<=HN_fst[i,2])&(all_info_separated[,3]>=HN_fst[i,2]))
if(length(k)>0){
print(paste0(i,"/",nrow(HN_fst)));
j<-c(j,i)
}
}
HN_fst2<-HN_fst[j,]#在cCRE中的fstpos
fimo_HN_fst<-NULL
for(i in 1:nrow(HN_fst2)){
inter_HN<-all_info_separated[which((all_info_separated[,1]==HN_fst2[i,1])&(all_info_separated[,2]<=HN_fst2[i,2])&(all_info_separated[,3]>=HN_fst2[i,2])),]
fimo_inrer<-NULL
for(j in 1:nrow(inter_HN)){
pos<-paste0("::",inter_HN[j,1],":",inter_HN[j,2],"-",inter_HN[j,3])
fimo_inrer2<-fimo_HN[which(fimo_HN[,3]==pos),]
fimo_inrer2$pos<-paste0(HN_fst2[i,1],":",HN_fst2[i,2])
fimo_inrer<-rbind(fimo_inrer,fimo_inrer2)
}
fimo_HN_fst<-rbind(fimo_HN_fst,fimo_inrer)
print(paste0(i,"/",nrow(HN_fst2)))
}
x1_fst<-unlist(strsplit(fimo_HN_fst$sequence_name, "::|:|-"))
x2_fst<-unlist(strsplit(fimo_HN_fst$pos, "::|:|-"))
fimo_HN_fst$chr1<-x1_fst[seq(1, length(x1_fst), by = 4)+1]
fimo_HN_fst$start1<-as.numeric(x1_fst[seq(1, length(x1_fst), by = 4)+2])+as.numeric(fimo_HN_fst$start)
fimo_HN_fst$stop1<-as.numeric(x1_fst[seq(1, length(x1_fst), by = 4)+2])+as.numeric(fimo_HN_fst$stop)
fimo_HN_fst$pos2<-x2_fst[seq(1, length(x2_fst), by = 2)]
fimo_HN_fst$pos3<-x2_fst[seq(1, length(x2_fst), by = 2)+1]
fimo_HN_fst$pd<-ifelse((fimo_HN_fst$pos2==fimo_HN_fst$chr1)&(fimo_HN_fst$pos3>fimo_HN_fst$start1)&(fimo_HN_fst$pos3<fimo_HN_fst$stop1),"Y","N")
####fimo_HN_fst_freq1<-fimo_HN_fst

###fimo_HN_fst[fimo_HN_fst$pd=="Y",]
fimo_HN_fst_x3<-fimo_HN_fst[fimo_HN_fst$pd=="Y",] ##海南牛高频SNP在fimo有TF富集的条目
}
############调控
RE_HN<-read.table("RE_HN.bed",header=F)
for(k in 1:nrow(fimo_HN_fst_x3)){
dex<-which(paste0("::",RE_HN$V1,":",RE_HN$V2,"-",RE_HN$V3)==fimo_HN_fst_x3$sequence_name[k])
fimo_HN_fst_x3$RE<-ifelse(length(dex)>0,RE_HN$V4[dex],"N")
print(paste0(k,"/",nrow(fimo_HN_fst_x3)))
}
############倒推
{##########HN-fst-RE。cluster
HN_fst<-read.table("1950.select.fst0.5.SNP.bed",header=T)
HN_fst$CHROM<-paste0("chr",HN_fst$CHROM)
#all_info_separated[k,]			set2[k,]
j<-c()
for(i in 1:nrow(HN_fst)){
k<-which((RE_HN[,1]==HN_fst[i,1])&(RE_HN[,2]<=HN_fst[i,2])&(RE_HN[,3]>=HN_fst[i,2]))
if(length(k)>0){
print(paste0(i,"/",nrow(HN_fst)));
j<-c(j,i)
}
}
HN_fst_RE<-HN_fst[j,]#在cCRE-REcluster中的fstpos
fimo_HN_fst_RE<-NULL
for(i in 1:nrow(HN_fst_RE)){
inter_HN<-RE_HN[which((RE_HN[,1]==HN_fst_RE[i,1])&(RE_HN[,2]<=HN_fst_RE[i,2])&(RE_HN[,3]>=HN_fst_RE[i,2])),]
fimo_inrer<-NULL
for(j in 1:nrow(inter_HN)){
pos<-paste0("::",inter_HN[j,1],":",inter_HN[j,2],"-",inter_HN[j,3]-1)
fimo_inrer2<-fimo_HN[which(fimo_HN[,3]==pos),]
fimo_inrer2$pos<-paste0(HN_fst_RE[i,1],":",HN_fst_RE[i,2])
fimo_inrer2$gene<-inter_HN[j,4]
fimo_inrer<-rbind(fimo_inrer,fimo_inrer2)
}
fimo_HN_fst_RE<-rbind(fimo_HN_fst_RE,fimo_inrer)
print(paste0(i,"/",nrow(HN_fst_RE)))
}

x1_fst<-unlist(strsplit(fimo_HN_fst_RE$sequence_name, "::|:|-"))
x2_fst<-unlist(strsplit(fimo_HN_fst_RE$pos, "::|:|-"))
fimo_HN_fst_RE$chr1<-x1_fst[seq(1, length(x1_fst), by = 4)+1]
fimo_HN_fst_RE$start1<-as.numeric(x1_fst[seq(1, length(x1_fst), by = 4)+2])+as.numeric(fimo_HN_fst_RE$start)
fimo_HN_fst_RE$stop1<-as.numeric(x1_fst[seq(1, length(x1_fst), by = 4)+2])+as.numeric(fimo_HN_fst_RE$stop)
fimo_HN_fst_RE$pos2<-x2_fst[seq(1, length(x2_fst), by = 2)]
fimo_HN_fst_RE$pos3<-x2_fst[seq(1, length(x2_fst), by = 2)+1]
fimo_HN_fst_RE$pd<-ifelse((fimo_HN_fst_RE$pos2==fimo_HN_fst_RE$chr1)&(fimo_HN_fst_RE$pos3>fimo_HN_fst_RE$start1)&(fimo_HN_fst_RE$pos3<fimo_HN_fst_RE$stop1),"Y","N")
####fimo_HN_fst_RE_freq1<-fimo_HN_fst_RE
xm<-paste0("::",HN_all_info_separated[,1],":",HN_all_info_separated[,2],"-",HN_all_info_separated[,3])
xmm<-HN_all_info_separated[match(fimo_HN_fst_RE$sequence_name,xm),]
fimo_HN_fst_RE$tissue<-xmm$organ
fimo_HN_fst_RE$celltype<-xmm$celltype
write.table(fimo_HN_fst_RE,"fimo_HN_fst_RE.txt",sep="\t",col.names = T,row.names = F,quote = FALSE)

###fimo_HN_fst_RE[fimo_HN_fst_RE$pd=="Y",]
fimo_HN_fst_RE_x3<-fimo_HN_fst_RE[fimo_HN_fst_RE$pd=="Y",] ##海南牛高频SNP在fimo有TF富集的条目
}

############倒推
RE_MG<-read.table("RE_MG.bed",header=F)
{##########MG-fst-RE。cluster
MG_fst<-read.table("1950.select.fst0.5.SNP.bed",header=T)
MG_fst$CHROM<-paste0("chr",MG_fst$CHROM)
#all_info_separated[k,]			set2[k,]
j<-c()
for(i in 1:nrow(MG_fst)){
k<-which((RE_MG[,1]==MG_fst[i,1])&(RE_MG[,2]<=MG_fst[i,2])&(RE_MG[,3]>=MG_fst[i,2]))
if(length(k)>0){
print(paste0(i,"/",nrow(MG_fst)));
j<-c(j,i)
}
}
MG_fst_RE<-MG_fst[j,]#在cCRE-REcluster中的fstpos
fimo_MG_fst_RE<-NULL
for(i in 1:nrow(MG_fst_RE)){
inter_MG<-RE_MG[which((RE_MG[,1]==MG_fst_RE[i,1])&(RE_MG[,2]<=MG_fst_RE[i,2])&(RE_MG[,3]>=MG_fst_RE[i,2])),]
fimo_inrer<-NULL
for(j in 1:nrow(inter_MG)){
pos<-paste0("::",inter_MG[j,1],":",inter_MG[j,2],"-",inter_MG[j,3]-1)
fimo_inrer2<-fimo_MG[which(fimo_MG[,3]==pos),]
fimo_inrer2$pos<-paste0(MG_fst_RE[i,1],":",MG_fst_RE[i,2])
fimo_inrer2$gene<-inter_MG[j,4]
fimo_inrer<-rbind(fimo_inrer,fimo_inrer2)
}
fimo_MG_fst_RE<-rbind(fimo_MG_fst_RE,fimo_inrer)
print(paste0(i,"/",nrow(MG_fst_RE)))
}

x1_fst<-unlist(strsplit(fimo_MG_fst_RE$sequence_name, "::|:|-"))
x2_fst<-unlist(strsplit(fimo_MG_fst_RE$pos, "::|:|-"))
fimo_MG_fst_RE$chr1<-x1_fst[seq(1, length(x1_fst), by = 4)+1]
fimo_MG_fst_RE$start1<-as.numeric(x1_fst[seq(1, length(x1_fst), by = 4)+2])+as.numeric(fimo_MG_fst_RE$start)
fimo_MG_fst_RE$stop1<-as.numeric(x1_fst[seq(1, length(x1_fst), by = 4)+2])+as.numeric(fimo_MG_fst_RE$stop)
fimo_MG_fst_RE$pos2<-x2_fst[seq(1, length(x2_fst), by = 2)]
fimo_MG_fst_RE$pos3<-x2_fst[seq(1, length(x2_fst), by = 2)+1]
fimo_MG_fst_RE$pd<-ifelse((fimo_MG_fst_RE$pos2==fimo_MG_fst_RE$chr1)&(fimo_MG_fst_RE$pos3>fimo_MG_fst_RE$start1)&(fimo_MG_fst_RE$pos3<fimo_MG_fst_RE$stop1),"Y","N")
####fimo_MG_fst_RE_freq1<-fimo_MG_fst_RE
ym<-paste0("::",MG_all_info_separated[,1],":",MG_all_info_separated[,2],"-",MG_all_info_separated[,3])
ymm<-MG_all_info_separated[match(fimo_MG_fst_RE$sequence_name,ym),]
fimo_MG_fst_RE$tissue<-ymm$organ
fimo_MG_fst_RE$celltype<-ymm$celltype
write.table(fimo_MG_fst_RE,"fimo_MG_fst_RE.txt",sep="\t",col.names = T,row.names = F,quote = FALSE)

###fimo_MG_fst_RE[fimo_MG_fst_RE$pd=="Y",]
fimo_MG_fst_RE_x3<-fimo_MG_fst_RE[fimo_MG_fst_RE$pd=="Y",] ##海南牛高频SNP在fimo有TF富集的条目
}



network<-rbind(HN_all_info_separated[,c(10:12)],MG_all_info_separated[,c(10:12)])
network2<-as.data.frame(table(network$V10,network$organ,network$celltype))
network2<-network2[!network2$Freq==0,]

network$Name <- ifelse(df$Name == "Tom", "Tommy", df$Name)
 To investigate whether some functional pathways evolve in a coordinated manner, we performed GO enrichment analysis on the genes near each group of the cCREs.
print(df)


#########################################################upset-merge
#conda activate scarlink-env
#R
HN_MG_int<-read.table("/storage/public/home/2021060195/SC/liftover/MG_HN_peak.bed",header=F)
subset_min <- HN_MG_int[, c(2, 5)]
# 提取第二列和第四列
subset_max <- HN_MG_int[, c(3, 6)]
# 计算第一列和第三列每一行的最小值
min_values <- apply(subset_min, 1, min)
# 计算第二列和第四列每一行的最大值
max_values <- apply(subset_max, 1, max)
# 创建新的矩阵，将每行的最小值和最大值合并
result_matrix <- cbind(HN_MG_int[,1],cbind(min_values, max_values))
write.table(result_matrix, file = "/storage/public/home/2021060195/SC/liftover/MG_HN_peak_expand.txt", sep = "\t", row.names = FALSE, col.names = FALSE, quote = FALSE)

library(data.table)
#HN_merge<-read.table("/storage/public/home/2021060195/SC/liftover/HN_all_peak_region_merge.bed",header=F)
#MG_merge<-read.table("/storage/public/home/2021060195/SC/liftover/MG_all_peak_region_merge.bed",header=F)
Peak_merge<-read.table("/storage/public/home/2021060195/SC/liftover/nonredundant_658706peak.bed",header=F)
for(i in c("HN")){
for(j in c("Hg38","Mm10","SusScr11","EquCab3","GCF_016772045.2")){
  #assign(paste(i, j, sep = "_"), read.table(paste(i,"_0.5_", j,".bed", sep = ""),header=F))
  assign(paste(i, j,"merge_unmap", sep = "_"), read.table(paste("/storage/public/home/2021060195/SC/liftover/50-merge-nonredundant/",i,"_unmap_0.5_", j,".bed", sep = ""),header=F))
}
}
Peak_merge_set<-paste("HN",Peak_merge[,1],Peak_merge[,2],Peak_merge[,3],sep = "-")
#HN_merge_set<-paste("HN",HN_merge[,1],HN_merge[,2],HN_merge[,3],sep = "-")
#MG_merge_set<-paste("MG",MG_merge[,1],MG_merge[,2],MG_merge[,3],sep = "-")
for(i in c("HN")){
for(j in c("Hg38","Mm10","SusScr11","EquCab3","GCF_016772045.2")){
  #assign(paste(i, j, sep = "_"), read.table(paste(i,"_0.5_", j,".bed", sep = ""),header=F))
  assign(paste(i, j,"merge_unmap_set", sep = "_"), paste(i,get(paste(i, j,"merge_unmap", sep = "_"))[,1],get(paste(i, j,"merge_unmap", sep = "_"))[,2],get(paste(i, j,"merge_unmap", sep = "_"))[,3],sep = "-"))
}
}

set3<-as.data.frame(list(name=c(Peak_merge_set)))
set3$hg[!Peak_merge_set%in%HN_Hg38_merge_unmap_set]<-"1"
set3$hg[Peak_merge_set%in%HN_Hg38_merge_unmap_set]<-"0"
set3$mm[!Peak_merge_set%in%HN_Mm10_merge_unmap_set]<-"1"
set3$mm[Peak_merge_set%in%HN_Mm10_merge_unmap_set]<-"0"
set3$pig[!Peak_merge_set%in%HN_SusScr11_merge_unmap_set]<-"1"
set3$pig[Peak_merge_set%in%HN_SusScr11_merge_unmap_set]<-"0"
set3$horse[!Peak_merge_set%in%HN_EquCab3_merge_unmap_set]<-"1"
set3$horse[Peak_merge_set%in%HN_EquCab3_merge_unmap_set]<-"0"
set3$sheep[!Peak_merge_set%in%HN_GCF_016772045.2_merge_unmap_set]<-"1"
set3$sheep[Peak_merge_set%in%HN_GCF_016772045.2_merge_unmap_set]<-"0"
set3[2:6]<-lapply(set3[2:6], as.numeric)

#cattle specific
cattle_specific<-Reduce(intersect,list(HN_Hg38_merge_unmap_set,HN_Mm10_merge_unmap_set,HN_SusScr11_merge_unmap_set,HN_EquCab3_merge_unmap_set,HN_GCF_016772045.2_merge_unmap_set))
library(UpSetR)
upsetplot2<-upset(set3, nsets = 4, mb.ratio = c(0.7, 0.3),sets=c("sheep","horse","pig","mm","hg"),keep.order = TRUE,order.by = c("freq", "degree"), decreasing = c(TRUE,TRUE),sets.bar.color = c("#20498D", "#20498D", "#20498D", "#20498D","#20498D"),matrix.color = '#225EA8',point.size = 3,line.size = 1,shade.color = 'grey',shade.alpha = 0.2,matrix.dot.alpha = 0.7)


#############near gene
#https://jokergoo.github.io/BioMartGOGeneSets/articles/supported_organisms.html
library(rtracklayer)
library(rGREAT)
library(ggplot2)
library(tidyverse)
gr_HN = import("set2_HN_merge.only.2.bed")
res_HN = great(gr_HN, "GO:BP", biomart_dataset = "btaurus_gene_ensembl")
tb_HN = getEnrichmentTable(res_HN)
tb_HN2<-getEnrichmentTable(res_HN) %>%
	tibble::as_tibble() %>%
	dplyr::select(id,genome_fraction,observed_region_hits,p_value) %>%
	dplyr::mutate(
		updated_pvalue = exp(pbinom(
			q = observed_region_hits - 1,
			size = attr(res_HN,which = "n_total"),
			prob = genome_fraction,
			lower.tail = F,
			log.p = T
		))
	)
tb_HN$p_value<-tb_HN2$updated_pvalue
tb_HN$log<-0
for(i in which(tb_HN$p_value==0)){
tb_HN$log[i]<-pbinom(q = tb_HN$observed_region_hits[i] - 1,size = attr(res_HN,which = "n_total"),prob = tb_HN$genome_fraction[i],lower.tail = F,log.p = T) / log(10)
}
tb_HN$p_adjust<-p.adjust(tb_HN$p_value,"BH")



gr_MG = import("set2_MG_merge.only.2.bed")
res_MG = great(gr_MG, "GO:BP", biomart_dataset = "btaurus_gene_ensembl")
tb_MG = getEnrichmentTable(res_MG)
tb_MG2<-getEnrichmentTable(res_MG) %>%
	tibble::as_tibble() %>%
	dplyr::select(id,genome_fraction,observed_region_hits,p_value) %>%
	dplyr::mutate(
		updated_pvalue = exp(pbinom(
			q = observed_region_hits - 1,
			size = attr(res_MG,which = "n_total"),
			prob = genome_fraction,
			lower.tail = F,
			log.p = T
		))
	)
tb_MG$p_value<-tb_MG2$updated_pvalue
tb_MG$log<-0
for(i in which(tb_MG$p_value==0)){
tb_MG$log[i]<-pbinom(q = tb_MG$observed_region_hits[i] - 1,size = attr(res_MG,which = "n_total"),prob = tb_MG$genome_fraction[i],lower.tail = F,log.p = T) / log(10)
}
tb_MG$p_adjust<-p.adjust(tb_MG$p_value,"BH")


gr = import("set2_HN_MG_merge.2.bed")
res = great(gr, "GO:BP", biomart_dataset = "btaurus_gene_ensembl")
tb = getEnrichmentTable(res)
tb2<-getEnrichmentTable(res) %>%
	tibble::as_tibble() %>%
	dplyr::select(id,genome_fraction,observed_region_hits,p_value) %>%
	dplyr::mutate(
		updated_pvalue = exp(pbinom(
			q = observed_region_hits - 1,
			size = attr(res,which = "n_total"),
			prob = genome_fraction,
			lower.tail = F,
			log.p = T
		))
	)
tb$p_value[which(tb$p_value==0)]<-tb2$updated_pvalue[tb2$p_value==0]
tb$p_adjust<-p.adjust(tb$p_value,"BH")

library(openxlsx)
wb <- createWorkbook()
addWorksheet(wb, "tb_all")
addWorksheet(wb, "tb_HN")
addWorksheet(wb, "tb_MG")
writeData(wb, sheet="tb_all", tb)
writeData(wb, sheet="tb_HN", tb_HN)
writeData(wb, sheet="tb_MG", tb_MG)
saveWorkbook(wb, file = "multiple_sheets.xlsx", overwrite = TRUE)

tb2<-tb[(tb$fold_enrichment>1.5)&(tb$p_adjust<2.2e-16),]
tb3<-tb2[order(tb2$p_adjust),]
tb4<-tb3[1:15,]
tb4 %>%
    mutate(description = fct_reorder(description, -log10(p_adjust))) %>%
    ggplot( aes(x=description, y=-log10(p_adjust))) +
    geom_bar(stat="identity", fill="#f68060") +
    coord_flip() +
    xlab("") +
    theme_bw()+theme(panel.grid = element_blank())
	
tb_HN2<-tb_HN[(tb_HN$fold_enrichment>1.5)&(tb_HN$p_adjust<2.2e-16),]
tb_HN3<-tb_HN2[order(tb_HN2$p_adjust),]
tb_HN4<-tb_HN3[1:15,]
tb_HN4 %>%
    mutate(description = fct_reorder(description, -log10(p_adjust))) %>%
    ggplot( aes(x=description, y=-log10(p_adjust))) +
    geom_bar(stat="identity", fill="#f68060") +
    coord_flip() +
    xlab("") +
    theme_bw()+theme(panel.grid = element_blank())
	
tb_MG2<-tb_MG[(tb_MG$fold_enrichment>1.5)&(tb_MG$p_adjust<0.05),]
tb_MG3<-tb_MG2[order(tb_MG2$p_adjust),]
tb_MG4<-tb_MG3[1:15,]
tb_MG4 %>%
    mutate(description = fct_reorder(description, -log10(p_adjust))) %>%
    ggplot( aes(x=description, y=-log10(p_adjust))) +
    geom_bar(stat="identity", fill="#f68060") +
    coord_flip() +
    xlab("") +
    theme_bw()+theme(panel.grid = element_blank())